### PR TITLE
https://github.com/javers/javers/issues/97 smare list compare

### PR DESCRIPTION
**Pull request by Kornel**
- Added a smart comparison for lists based on the Levenshtein
  edit distance problem, see package-info for brief description

### CODE REVIEW

**Comment by pawelszymczyk** [comment]
I have prepared some performance tests for this appender, it was compared with old (Dummy) appender. There are two types of tests, results are below.

Tests have been running on :
2,3 GHz Intel Core i7
16 GB 1600 MHz DDR3

Source code:
https://github.com/javers/javers/blob/%2397_smart_list_compare_perf/javers-persistence-sql/src/test/groovy/org/javers/repository/sql/ListsAppenderPerformanceTest.groovy

## In each iteration was changed all elements in list

1. Unknown db

 | appender  | elements | commits | total time [ms]  |
 | --------- | -------- | ------- | ------------ |
 | Dummy | 1000  | 100  | 138.00  | 
 | Dummy | 5000  | 100  | 497.00  | 
 | Dummy | 10000  | 100  | 1029.00  | 
 | Dummy | 1000  | 1000  | 1604.00  | 
 | Dummy | 5000  | 1000  | 5321.00  | 
 | Dummy | 10000  | 1000  | 13369.00  | 
 | Levenstein |  1000  | 100  | 1371.00  | 
 | Levenstein |  5000  | 100  | 32722.00  | 
 | Levenstein |  10000  | 100  | 151449.00  | 
 | Levenstein |  1000  | 1000  | 14222.00  | 
 | Levenstein |  5000  | 1000  | 336995.00  | 
 | Levenstein |  10000  | 1000  | 1765647.00  | 


2. h2

 | appender  | elements | commits | total time [ms]  |
 | --------- | -------- | ------- | ------------ |  
 | Dummy | 200  | 100  | 92.00  | 
 | Dummy | 400  | 100  | 91.00  | 
 | Dummy | 600  | 100  | 101.00  | 
 | Dummy | 800  | 100  | 120.00  | 
 | Dummy | 1000  | 100  | 160.00  | 
 | Dummy | 3000  | 100  | 343.00  | 
 | Dummy | 5000  | 100  | 544.00  | 
 | Dummy | 200  | 1000  | 377.00  | 
 | Dummy | 400  | 1000  | 487.00  | 
 | Dummy | 600  | 1000  | 678.00  | 
 | Dummy | 800  | 1000  | 824.00  | 
 | Dummy | 1000  | 1000  | 1498.00  | 
 | Dummy | 3000  | 1000  | 3921.00  | 
 | Dummy | 5000  | 1000  | 7552.00  | 
 | Levenstein |  200  | 100  | 72.00  | 
 | Levenstein |  400  | 100  | 242.00  | 
 | Levenstein |  600  | 100  | 591.00  | 
 | Levenstein |  800  | 100  | 1193.00  | 
 | Levenstein |  1000  | 100  | 1552.00  | 
 | Levenstein |  3000  | 100  | 12509.00  | 
 | Levenstein |  5000  | 100  | 33834.00  | 
 | Levenstein |  200  | 1000  | 728.00  | 
 | Levenstein |  400  | 1000  | 2487.00  | 
 | Levenstein |  600  | 1000  | 5358.00  | 
 | Levenstein |  800  | 1000  | 9348.00  | 
 | Levenstein |  1000  | 1000  | 14146.00  | 
 | Levenstein |  3000  | 1000  | 121817.00  | 
 | Levenstein |  5000  | 1000  | 343028.00  | 

3. h2

 | appender  | elements | commits | total time [ms]  |
 | --------- | -------- | ------- | ------------ |
 | Dummy | 200  | 100  | 94.00  | 0.94
 | Dummy | 400  | 100  | 99.00  | 0.99
 | Dummy | 600  | 100  | 108.00  | 1.08
 | Dummy | 800  | 100  | 153.00  | 1.53
 | Dummy | 1000  | 100  | 163.00  | 1.63
 | Dummy | 3000  | 100  | 357.00  | 3.57
 | Dummy | 5000  | 100  | 553.00  | 5.53
 | Dummy | 200  | 1000  | 384.00  | 0.384
 | Dummy | 400  | 1000  | 504.00  | 0.504
 | Dummy | 600  | 1000  | 711.00  | 0.711
 | Dummy | 800  | 1000  | 876.00  | 0.876
 | Dummy | 1000  | 1000  | 1671.00  | 1.671
 | Dummy | 3000  | 1000  | 3796.00  | 3.796
 | Dummy | 5000  | 1000  | 7725.00  | 7.725
 | Levenstein |  200  | 100  | 71.00  | 0.71
 | Levenstein |  400  | 100  | 232.00  | 2.32
 | Levenstein |  600  | 100  | 578.00  | 5.78
 | Levenstein |  800  | 100  | 958.00  | 9.58
 | Levenstein |  1000  | 100  | 1403.00  | 14.03
 | Levenstein |  3000  | 100  | 12595.00  | 125.95
 | Levenstein |  5000  | 100  | 34554.00  | 345.54
 | Levenstein |  200  | 1000  | 749.00  | 0.749
 | Levenstein |  400  | 1000  | 2506.00  | 2.506
 | Levenstein |  600  | 1000  | 5318.00  | 5.318
 | Levenstein |  800  | 1000  | 9280.00  | 9.28
 | Levenstein |  1000  | 1000  | 14166.00  | 14.166
 | Levenstein |  3000  | 1000  | 125769.00  | 125.769
 | Levenstein |  5000  | 1000  | 350059.00  | 350.059


## In each iteration was removed tenth element in list

1. psql

| appender  | elements | commits | total time [ms]  |
| --------- | -------- | ------- | ------------ |
| Dummy     | 1050     | 100     | 94.00        |
| Dummy     | 5050     | 100     | 454.00       |
| Dummy     | 10050    | 100     | 875.00       |
| Dummy     | 1050     | 1000    | 608.00       |
| Dummy| 5050 | 1000 | 5205.00 |
| Dummy| 10050 | 1000 | 11012.00 |
| Levenstein| 1050| 100 | 1385.00 |
| Levenstein| 5050| 100 | 33533.00 |
| Levenstein| 10050| 100 | 153643.00 |
| Levenstein| 1050| 1000 | 6351.00 |
| Levenstein| 5050| 1000 | 277437.00 |
| Levenstein| 10050| 1000 | 1420944.00 |


2. psql

| appender  | elements | commits | total time [ms]  |
| --------- | -------- | ------- | ------ |
| Dummy| 1050| 100 | 136.00 | 1.36
| Dummy| 5050| 100 | 484.00 | 4.84
| Dummy| 10050| 100 | 997.00 | 9.97
| Dummy| 1050| 1000 | 715.00 | 0.715
| Dummy| 5050| 1000 | 6041.00 | 6.041
| Levenstein| 1050| 100 | 941.00 | 9.41
| Levenstein| 5050| 100 | 19926.00 | 199.26
| Levenstein| 10050| 100 | 101740.00 | 1017.4
| Levenstein| 1050| 1000 | 3579.00 | 3.579
| Levenstein| 5050| 1000 | 176392.00 | 176.392

3. h2

| appender  | elements | commits | total time [ms]  |
| --------- | -------- | ------- | ------ |
| Dummy| 150| 100 | 116.00 | 1.16
| Dummy| 250| 100 | 65.00 | 0.65
| Dummy| 550| 100 | 95.00 | 0.95
| Dummy| 1050| 100 | 143.00 | 1.43
| Dummy| 5050| 100 | 570.00 | 5.7
| Dummy| 10050| 100 | 1682.00 | 16.82
| Dummy| 1050| 1000 | 668.00 | 0.668
| Dummy| 5050| 1000 | 5317.00 | 5.317
| Levenstein| 150| 100 | 116.00 | 1.16
| Levenstein| 250| 100 | 87.00 | 0.87
| Levenstein| 550| 100 | 269.00 | 2.69
| Levenstein| 1050| 100 | 958.00 | 9.58
| Levenstein| 5050| 100 | 19916.00 | 199.16
| Levenstein| 10050| 100 | 101618.00 | 1016.18
| Levenstein| 1050| 1000 | 3853.00 | 3.853
| Levenstein| 5050| 1000 | 179340.00 | 179.34


4. h2

| appender  | elements | commits | total time [ms]  |
| --------- | -------- | ------- | ------ |
| Dummy| 200| 100 | 122.00 | 1.22
| Dummy| 400| 100 | 80.00 | 0.8
| Dummy| 600| 100 | 93.00 | 0.93
| Dummy| 800| 100 | 124.00 | 1.24
| Dummy| 1000| 100 | 154.00 | 1.54
| Dummy| 3000| 100 | 319.00 | 3.19
| Dummy| 5000| 100 | 513.00 | 5.13
| Levenstein| 200| 100 | 80.00 | 0.8
| Levenstein| 400| 100 | 203.00 | 2.03
| Levenstein| 600| 100 | 469.00 | 4.69
| Levenstein| 800| 100 | 817.00 | 8.17
| Levenstein| 1000| 100 | 1272.00 | 12.72
| Levenstein| 3000| 100 | 11679.00 | 116.79
| Levenstein| 5000| 100 | 33317.00 | 333.17

**Comment by pawelszymczyk** [comment]
I've just added method that allow to register LevenshteinListChangeAppender in Javers, the default is ListChangeAppender.

**Comment by pawelszymczyk** [comment]
@bartoszwalacik review pls

File: javers-core/src/main/java/org/javers/core/diff/appenders/ListChangeAppender.java [hunk 1]
```diff
@@ -18,9 +19,7 @@
 /**
  * @author pawel szymczyk
  */
-class ListChangeAppender extends CorePropertyChangeAppender<ListChange> {
-
-    private static final Logger logger = LoggerFactory.getLogger(ListChangeAppender.class);
+public class ListChangeAppender extends CorePropertyChangeAppender<ListChange> {
```
Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/core/diff/appenders/ListChangeAppender.java]: ale dlaczego public ...?

Comment (bartoszwalacik) [inline javers-core/src/main/java/org/javers/core/diff/appenders/ListChangeAppender.java]: ok widzę, to przez tego enuma :)

**Comment by bartoszwalacik** [comment]
@Kornel , there is a bug in the algorithm.
no add/remove elements are detected at the beginning of the list

I've committed a failing test to cover this

**Comment by Kornel** [comment]
nice catch, here comes the fix :)

**Pull request merged**